kv/kvserver: TestLeaseTransferRejectedIfTargetNeedsSnapshot failed

kv/kvserver.TestLeaseTransferRejectedIfTargetNeedsSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7596751?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7596751?buildTab=artifacts#/) on release-22.2 @ [022d19ab6c1c6f6dd477d5120b4c47a46f744238](https://github.com/cockroachdb/cockroach/commits/022d19ab6c1c6f6dd477d5120b4c47a46f744238):


```
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot3692090459
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestLeaseTransferRejectedIfTargetNeedsSnapshot
    client_replica_test.go:3140: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot3692090459
--- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot (7.08s)
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=false
    client_replica_test.go:3138: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2994/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/client_replica_test.go:3138
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2994/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/subtest.go:23
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 3
        	Test:       	TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=false
    --- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=false (3.65s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseTransferRejectedIfTargetNeedsSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21631

kv/kvserver.TestLeaseTransferRejectedIfTargetNeedsSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7688946?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7688946?buildTab=artifacts#/) on release-22.2 @ [fd133ba3af3954c93584a501cf68dfa3131e586c](https://github.com/cockroachdb/cockroach/commits/fd133ba3af3954c93584a501cf68dfa3131e586c):


```
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot4030135416
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestLeaseTransferRejectedIfTargetNeedsSnapshot
    client_replica_test.go:3140: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot4030135416
--- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot (6.50s)
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true
    client_replica_test.go:3138: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3251/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/client_replica_test.go:3138
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3251/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/subtest.go:23
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 4
        	Test:       	TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true
    --- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true (2.65s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseTransferRejectedIfTargetNeedsSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestLeaseTransferRejectedIfTargetNeedsSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7767054?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7767054?buildTab=artifacts#/) on release-22.2 @ [d978988afcfc78c02c23a3836ba90e4a4faeeb25](https://github.com/cockroachdb/cockroach/commits/d978988afcfc78c02c23a3836ba90e4a4faeeb25):


```
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot467253111
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestLeaseTransferRejectedIfTargetNeedsSnapshot
    client_replica_test.go:3140: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaseTransferRejectedIfTargetNeedsSnapshot467253111
--- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot (8.41s)
=== RUN   TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
    client_replica_test.go:3138: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3231/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/client_replica_test.go:3138
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3231/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/subtest.go:23
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 4
        	Test:       	TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true
    --- FAIL: TestLeaseTransferRejectedIfTargetNeedsSnapshot/reject-after-revoke=true (4.11s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseTransferRejectedIfTargetNeedsSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -3136,11 +3136,14 @@ func TestLeaseTransferRejectedIfTargetNeedsSnapshot(t *testing.T) {
 		// the lease transfer rejection came after the previous lease was revoked,
 		// then node 0 must have re-acquired the lease (with a new sequence number)
 		// in order to transfer it to node 2.
+		// NB: we use LessOrEqual and not Equal to avoid flakiness if the lease is
+		// lost and reacquired multiple times. This assertion is not the focus of
+		// the test.
 		expSeq := preLease.Sequence + 1
 		if rejectAfterRevoke {
 			expSeq++
 		}
-		require.Equal(t, expSeq, postLease.Sequence)
+		require.LessOrEqual(t, expSeq, postLease.Sequence)
 	})
 }
 

